include/cuco/detail/traits.hpp

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/extent_base.cuh>
+#include <cuco/detail/probing_scheme_base.cuh>
+// #include <cuco/detail/storage/storage_base.cuh> // FIXME check inheritance
+#include <cuco/storage.cuh>
+
+#include <memory>
+#include <type_traits>
+
+namespace cuco::experimental::detail {
+
+// Trait to check if T is a valid probing scheme type
+template <typename T, typename Enable = void>
+struct is_probing_scheme : std::false_type {
+};
+
+template <typename T>
+struct is_probing_scheme<
+  T,
+  std::enable_if_t<
+    std::is_base_of_v<cuco::experimental::detail::probing_scheme_base<T::cg_size>, T>>>
+  : std::true_type {
+};
+
+template <typename T>
+inline constexpr bool is_probing_scheme_v = is_probing_scheme<T>::value;
+
+template <typename T, typename Enable = void>
+struct is_storage : std::false_type {
+};
+
+template <typename T>
+struct is_storage<
+  T,
+  std::enable_if_t<std::is_same_v<cuco::experimental::aow_storage<T::window_size>, T>>>
+  : std::true_type {
+};
+
+/* FIXME
+std::enable_if_t<std::is_base_of_v<
+  typename cuco::experimental::detail::storage_base<typename T::extent_type>,
+  T>>> : std::true_type {};
+*/
+
+template <typename T>
+inline constexpr bool is_storage_v = is_storage<T>::value;
+
+// Trait to check if T is a `cuco::extent`
+template <typename T, typename Enable = void>
+struct is_extent : std::false_type {
+};
+
+template <typename T>
+struct is_extent<
+  T,
+  std::enable_if_t<
+    std::is_base_of_v<typename cuco::experimental::detail::extent_base<typename T::value_type>, T>>>
+  : std::true_type {
+};
+
+template <typename T>
+inline constexpr bool is_extent_v = is_extent<T>::value;
+
+// Trait to check if T is allocator-like
+template <typename T, typename Enable = std::void_t<>>
+struct is_allocator : std::false_type {
+};
+
+template <typename T>
+struct is_allocator<T,
+                    std::void_t<typename T::value_type,
+                                decltype(std::declval<T&>().allocate(std::size_t{})),
+                                decltype(std::declval<T&>().deallocate(
+                                  std::declval<typename T::value_type*>(), std::size_t{}))>>
+  : std::true_type {
+};
+
+template <typename T>
+inline constexpr bool is_allocator_v = is_allocator<T>::value;
+
+template <typename Key>
+struct key_equal_traits {
+  template <typename T, typename = std::void_t<>>
+  struct is_equal_functor : std::false_type {
+  };
+
+  template <typename T>
+  struct is_equal_functor<
+    T,
+    std::void_t<std::enable_if_t<std::is_invocable_r_v<bool, T, Key const&, Key const&>>>>
+    : std::true_type {
+  };
+
+  template <typename T>
+  using is_equal_functor_t = typename is_equal_functor<T>::type;
+
+  template <typename T>
+  static constexpr bool is_equal_functor_v = is_equal_functor<T>::value;
+};
```

Comment:
This was a bit tricky. I needed a trait, which checks if a given user type `T` is a binary functor `Key, Key -> bool`. However, the `find_arg` function only knows of `T`, and does not directly know of `Key`. That's why this is a nested trait which can be passed as `find_arg<key_equal_traits<Key>::template is_equal_functor_t>(...)`.
It works, but I'm not proud of it. So I'm very open to suggestions to make this more elegant to use.